elements/lisk-utils/src/math/q.ts

```diff
@@ -0,0 +1,153 @@
+ * propagated, or distributed except according to the terms contained in the
+ * LICENSE file.
+ *
+ * Removal or modification of this copyright notice is prohibited.
+ */
+
+import { MAX_FRAC, ONE, Q_OPERATION, TWO } from './constants';
+
+const numberToQ = (base: bigint, val: number): bigint => {
+	const denominator = TWO ** base;
+	if (Number.isInteger(val)) {
+		return BigInt(val) * denominator;
+	}
+	const [intStr, fractionalStr] = val.toString().split('.');
+	return (
+		BigInt(intStr) * denominator +
+		(BigInt(fractionalStr) * denominator) / BigInt(10) ** BigInt(fractionalStr.length)
+	);
+};
+
+export const q = (val: number | bigint | Buffer, base: number | bigint): Q =>
+	Q.fromValue(val, base);
+
+export class Q {
+	private readonly _base: bigint;
+	private readonly _val: bigint;
+
+	public constructor(val: bigint, base: number | bigint) {
+		this._base = BigInt(base);
+		if (this._base > MAX_FRAC) {
+			throw new Error('Base exceeds max fraction allowed.');
+		}
+		this._val = val;
+		this._base = BigInt(base);
+	}
+
+	public static fromValue(val: number | bigint | Buffer | Q, base: number | bigint) {
+		if (val instanceof Q) {
+			return new Q(val._val, val._base);
+		}
+		const bigintBase = BigInt(base);
+		if (typeof val === 'number') {
+			return new Q(numberToQ(bigintBase, val), bigintBase);
+		}
+		if (typeof val === 'bigint') {
+			return new Q(val * TWO ** bigintBase, bigintBase);
+		}
+		return new Q(BigInt(`0x${val.toString('hex')}`), bigintBase);
+	}
+
+	public add(n: Q) {
+		if (this._base !== n._base) {
+			throw new Error('Base of input does not match.');
+		}
+		return new Q(this._val + n._val, this._base);
+	}
+
+	public sub(n: Q) {
+		if (this._base !== n._base) {
+			throw new Error('Invalid input, base does not match.');
+		}
+		if (this._val < n._val) {
+			throw new Error('Invalid input, output cannot be negative.');
+		}
+		return new Q(this._val - n._val, this._base);
+	}
+
+	public mul(n: Q) {
+		if (this._base !== n._base) {
+			throw new Error('Invalid input, base does not match.');
+		}
+		return new Q((this._val * n._val) >> this._base, this._base);
+	}
+
+	public div(n: Q) {
+		if (this._base !== n._base) {
+			throw new Error('Invalid input, base does not match.');
+		}
+		return new Q((this._val << this._base) / n._val, this._base);
+	}
+
+	public muldiv(n: Q, m: Q) {
+		if (this._base !== n._base || this._base !== m._base) {
+			throw new Error('Invalid input, base does not match.');
+		}
+		const x = this._val * n._val;
+		const y = x << this._base;
+		const z = y / m._val;
+
+		return new Q(this._roundDown(z), this._base);
+	}
+
+	public inv() {
+		const numerator = new Q(ONE << this._base, this._base);
+		return numerator.div(this);
+	}
+
+	public toInt(operation: Q_OPERATION = Q_OPERATION.ROUND_DOWN) {
+		if (operation === Q_OPERATION.ROUND_DOWN) {
+			return this._roundDown(this._val);
+		}
+		return this._roundUp(this._val);
```

Comment:
This looks awkward.

I don't think we need `toInt()` method. I propose to be consistent with JavaScript math functions and have `ceil()` and `floor()` instead. That could be what `_roundDown()` and `roundUp()` are all about.